vsock-bridge/src/socket.cpp

```diff
@@ -17,142 +17,86 @@ namespace vsockio
-				close();
-			}
-			else if (!_peer->queueEmpty())
-			{
-				// Peer has some queued data they never received
-				// Assuming this data is critical for the protocol, it should be ok to abort the connection straight away
-				Logger::instance->Log(Logger::DEBUG, "[socket] writeToOutput detected input peer is closed while having data remaining, closing (fd=", _fd, ")");
-				close();
-			}
-		}
-
-		return _sendQueue.empty();
+		return canSendModeData;
 	}
 
-	void Socket::queue(std::unique_ptr<Buffer>&& buffer)
+	bool Socket::read(Buffer& buffer)
 	{
-		_sendQueue.enqueue(std::move(buffer));
-
-		// to simplify logic we allow only 1 buffer for socket sinks
-		_queueFull = true;
-	}
-
-	std::unique_ptr<Buffer> Socket::read()
-	{
-		std::unique_ptr<Buffer> buffer{ BufferManager::getBuffer() };
-
-		while (true)
-		{
-			const int bytesRead = _impl.read(_fd, buffer->tail(), buffer->remainingCapacity());
-			int err = 0;
-			if (bytesRead > 0)
-			{
-				// New content read
-				// update byte count and enlarge buffer if needed
-
-				//Logger::instance->Log(Logger::DEBUG, "[socket] read returns ", bytesRead, " (fd=", _fd, ")");
-				buffer->produce(bytesRead);
-				if (!buffer->ensureCapacity())
-				{
-					break;
-				}
-			}
-			else if (bytesRead == 0)
-			{
-				// Source closed
-
-				Logger::instance->Log(Logger::DEBUG, "[socket] read returns 0, closing input (fd=", _fd, ")");
-				closeInput();
-				break;
-			}
-			else if ((err = errno) == EAGAIN || err == EWOULDBLOCK)
-			{
-				// No new data
-
-				_inputReady = false;
-				break;
-			}
-			else
-			{
-				// Error
-
-				Logger::instance->Log(Logger::WARNING, "[socket] error on read, closing input (fd=", _fd, "): ", strerror(err));
-				closeInput();
-				break;
-			}
-		}
-
-		return buffer;
+        if (!buffer.hasRemainingCapacity()) return false;
+
+        PERF_LOG("read");
+        const int bytesRead = _impl.read(_fd, buffer.tail(), buffer.remainingCapacity());
+        int err = 0;
+        if (bytesRead > 0)
+        {
+            // New content read
+
+            //Logger::instance->Log(Logger::DEBUG, "[socket] read returns ", bytesRead, " (fd=", _fd, ")");
+            buffer.produce(bytesRead);
+            return true;
+        }
+        else if (bytesRead == 0)
+        {
+            // Source closed
+
+            Logger::instance->Log(Logger::DEBUG, "[socket] read returns 0, closing (fd=", _fd, ")");
+            close();
+            return false;
+        }
+        else if ((err = errno) == EAGAIN || err == EWOULDBLOCK)
+        {
+            // No new data
+
+            return false;
+        }
+        else
+        {
+            // Error
+
+            Logger::instance->Log(Logger::WARNING, "[socket] error on read, closing (fd=", _fd, "): ", err, ", ", strerror(err));
+            close();
+            return false;
+        }
 	}
 
-	void Socket::send(Buffer& buffer)
+	bool Socket::send(Buffer& buffer)
 	{
+        bool canSendModeData = false;
```

Comment:
typo: "Mode" -> "More".

But reading the method twice, there's only one case we need `canSendModeData` here: when buffer is already consumed entering into the method, where we would return `false`. But this seems to be an exceptional case.

Should we remove this variable and return true at the end of the method?

Reply:
Removed the variable.